src/eva/vision/data/wsi/patching/samplers.py

```diff
@@ -93,13 +95,83 @@ def sample(
         y_range = range(0, layer_shape[1] - height, height - self.overlap[1])
         x_y = [(x, y) for x in x_range for y in y_range]
 
-        shuffled_indices = (
-            np.random.choice(len(x_y), self.max_samples, replace=False)
-            if self.max_samples
-            else range(len(x_y))
-        )
-        for i in shuffled_indices:
-            yield x_y[i]
+        indices = list(range(len(x_y)))
+        np.random.shuffle(indices)
+
+        if self.max_samples is not None and not ignore_max_samples:
+            for i in indices[: self.max_samples]:
+                yield x_y[i]
+        else:
+            for i in indices:
+                yield x_y[i]
+
+
+class ForegroundGridSampler(GridSampler):
+    """Sample patches based on a grid, only returning patches containing foreground.
+    
+    Args:
+        max_samples: The maximum number of samples to return.
+    """
+    def __init__(
+        self,
+        max_samples: int = 20,
+    ):
+        super().__init__(max_samples=max_samples)
+
+    def sample(
+        self,
+        width: int,
+        height: int,
+        layer_shape: tuple[int, int],
+        mask: np.ndarray,
+        mask_scale_factor: float,
+    ):
+        """Sample patches from a grid containing foreground.
+        
+        Args:
+            width: The width of the patches.
+            height: The height of the patches.
+            layer_shape: The shape of the layer.
+            mask: The mask of the image.
+            mask_scale_factor: The scale factor of the mask.
+        """
+        count = 0
+        for x, y in super().sample(width, height, layer_shape, ignore_max_samples=True):
+            if count >= self.max_samples:
+                break
+            if self.is_foreground(
+                mask, x, y, width, height, mask_scale_factor
+            ):
+                count += 1
+                yield x, y
+
+    def is_foreground(
+        self,
+        mask: np.ndarray,
+        x: int,
+        y: int,
+        width: int,
+        height: int,
+        mask_scale_factor: float,
```

Comment:
not sure if this is worth it: it would mean (1) resize the mask (linear interpolation) + (2) computing the foreground ratio for a much larger patch. Basically extra computation just for nicer code.(?)

Reply:
Good points, let's leave it as is